Realm/RLMSyncSessionRefreshHandle.mm

```diff
@@ -33,40 +33,132 @@ @interface RLMSyncSessionRefreshHandle () {
+                                               repeats:NO];
+        [[NSRunLoop currentRunLoop] addTimer:t forMode:NSDefaultRunLoopMode];
+        self.timer = t;
+    });
+}
+
+- (BOOL)onRefreshCompletionWithError:(NSError *)error json:(NSDictionary *)json {
+    RLMSyncUser *user = self.user;
+    if (!user) {
+        return NO;
+    }
+    if (json && !error) {
+        RLMAuthResponseModel *model = [[RLMAuthResponseModel alloc] initWithDictionary:json
+                                                                    requireAccessToken:YES
+                                                                   requireRefreshToken:NO];
+        if (!model) {
+            // Malformed JSON
+            error = [NSError errorWithDomain:RLMSyncErrorDomain
+                                        code:RLMSyncErrorBadResponse
+                                    userInfo:@{kRLMSyncErrorJSONKey: json}];
+            [user _unregisterRefreshHandleForURLPath:self.pathToRealm];
+            [self.timer invalidate];
+            [[RLMSyncManager sharedManager] _fireError:error];
+            return NO;
+        }
+
+        // Success
+        if (auto session = _session.lock()) {
+            if (session->state() != SyncSession::PublicState::Error) {
+                session->refresh_access_token([model.accessToken.token UTF8String], none);
+                NSDate *expiration = [NSDate dateWithTimeIntervalSince1970:model.accessToken.tokenData.expires];
+                [self scheduleRefreshTimer:expiration];
+                return YES;
+            }
+        }
+        // The session is dead or in a fatal error state.
+        [user _unregisterRefreshHandleForURLPath:self.pathToRealm];
+        [self.timer invalidate];
+        return NO;
+    }
+
+    // Something else went wrong
+    NSError *syncError = [NSError errorWithDomain:RLMSyncErrorDomain
+                                             code:RLMSyncErrorBadResponse
+                                         userInfo:@{kRLMSyncUnderlyingErrorKey: error}];
+    [[RLMSyncManager sharedManager] _fireError:syncError];
+    NSDate *nextFireDate = nil;
+    // Certain errors should trigger a retry.
+    if (error.domain == NSURLErrorDomain) {
+        switch (error.code) {
+            case NSURLErrorCannotConnectToHost:
+                // FIXME: 120 seconds is an arbitrarily chosen value, consider rationalizing it.
+                nextFireDate = [NSDate dateWithTimeIntervalSinceNow:120];
+                break;
+            case NSURLErrorNotConnectedToInternet:
+            case NSURLErrorNetworkConnectionLost:
+            case NSURLErrorTimedOut:
+            case NSURLErrorDNSLookupFailed:
+            case NSURLErrorCannotFindHost:
+                // FIXME: 30 seconds is an arbitrarily chosen value, consider rationalizing it.
+                nextFireDate = [NSDate dateWithTimeIntervalSinceNow:30];
+                break;
+            default:
+                break;
```

Comment:
The only errors that can occur are network errors, or errors I consider fatal (e.g. malformed JSON response, internal server error). That being said, my follow-up PR has some error recovery code.